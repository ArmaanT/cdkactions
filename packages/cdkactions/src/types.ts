/**
 * A generic string to string map.
 */
export interface StringMap { readonly [key: string]: string }


/**
 * Configuration for a shell environment.
 */
export interface RunProps {
  /**
   * The shell to use.
   */
  readonly shell?: 'bash' | 'pwsh' | 'python' | 'sh' | 'cmd' | 'powershell';

  /**
   * A custom working directory.
   */
  readonly workingDirectory?: string;
}


/**
 * A defaults configuration block.
 */
export interface DefaultsProps {
  /**
   * A RunProps block
   */
  readonly run?: RunProps;

  /**
   * The rest of the defaults block can be
   * any string key with values of any type.
   */
  readonly [key: string]: any;
}

/**
 * Permissions to apply to all scopes
 */
export type PermissionAccessAll = 'read-all' | 'write-all';

/**
 * Permissions to apply to individual scopes
 */
export type PermissionAccess = 'read' | 'write' | 'none';

/**
 * All available permission scopes
 */
export type PermissionScope = 'actions' | 'checks' | 'contents' | 'deployments' | 'id-token' | 'issues' | 'discussions' | 'packages' | 'pages' | 'pull-requests' | 'repository-projects' | 'security-events' | 'statuses';

/**
 * Configuration for permissions
 */
export type PermissionsProps = PermissionAccessAll | Partial<Record<PermissionScope, PermissionAccess>>;

