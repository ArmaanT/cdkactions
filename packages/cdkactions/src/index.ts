export * from './app';
export * from './job';
export * from './stack';
export * from './types';
export * from './workflow';
export * from './library';
