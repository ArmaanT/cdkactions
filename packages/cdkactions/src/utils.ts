import { StringMap } from './types';

/**
 * Fix the raw yaml returned by yaml.safeDump by removing quotes surrounding GH Action
 * expressions, removing double single quotes from within a GH Action expression,
 * and finally remove quotes from the top-level "on" key
 * @param rawYaml Raw yaml to be formatted
 */
export const formatGHActionsYaml = (rawYaml: string) => {
  // Remove quotes from GH Action expressions: ${{ ... }}
  const expression_re = /'(\${{ .* }})'/g;
  const fixedExpression = rawYaml.replace(expression_re, '$1');

  // Remove double single quotes from inside GH Action expressions:
  const double_re = /\${{ .*''.* }}/g;
  const removeDoubleQuote = (match: string, _offset: string, _string: string) => {
    return match.replace(/''/g, '\'');
  };
  const fixedDoubleQuotes = fixedExpression.replace(double_re, removeDoubleQuote);

  // Remove quotes from the top-level "on" key
  const on_re = /'on':/g;
  const fixedOn = fixedDoubleQuotes.replace(on_re, 'on:');

  // Removes nested single quotes like in "key: '${{ hashFiles('example/yarn.lock') }}'"
  const nested_re = /: '([^'\n\r]*'[^'\n\r]*'.*)'\n/g;
  const fixedNested = fixedOn.replace(nested_re, ': $1\n');

  return fixedNested;
};

/**
 * A helper function to recursively rename keys within an object
 * @param obj Object to rename
 * @param newKeys A dictionary of old to new key names
 */
export const renameKeys = (obj: any, newKeys: StringMap) => {
  if (typeof obj !== 'object') {
    return obj;
  }
  const keyValues = Object.keys(obj).map(key => {
    const newKey = newKeys[key] || key;
    const oldValue = obj[key];
    let newValue = oldValue;
    if (Array.isArray(oldValue)) {
      newValue = oldValue.map(item => renameKeys(item, newKeys));
    } else if (typeof oldValue === 'object') {
      newValue = renameKeys(oldValue, newKeys);
    }
    return { [newKey]: newValue };
  });
  return Object.assign({}, ...keyValues);
};

export const camelToSnake = (str: string) => str.replace(/[A-Z]/g, (group) => `_${group.toLowerCase()}`);
