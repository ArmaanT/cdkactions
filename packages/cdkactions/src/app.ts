import * as fs from 'fs';
import { Construct, Node } from 'constructs';
import { CDKActionsStack } from './library';

/**
 * Configuration for a cdkactions app.
 */
export interface AppProps {
  /**
   * The directory to synthesize GitHub Action manifests in.
   *
   * @default - "../workflows"
   */
  readonly outdir?: string;

  /**
   * Create a validation workflow to ensure cdkactions manifests are up to date.
   * @default - true
   */
  readonly createValidateWorkflow?: boolean;

  /**
   * Push updated manifests if they are out of date. Only used if `createValidateWorkflow` is true.
   * @default - false
   */
  readonly pushUpdatedManifests?: boolean;
}

/**
 * Represents a cdkactions application.
 */
export class App extends Construct {
  /**
   * The directory to synthesize GitHub Action manifests in.
   */
  public readonly outdir: string;

  /**
   * Defines a cdkaction app.
   * @param options Configuration options.
   */
  constructor(options?: AppProps) {
    // We unfortunately can't use Partial<AppProps> as a type for options due to JSII limitations.
    const config: AppProps = {
      createValidateWorkflow: true,
      pushUpdatedManifests: false,
      ...options,
    };

    super(undefined as any, '');
    this.outdir = config.outdir ?? '../workflows';
    if (config.createValidateWorkflow) {
      new CDKActionsStack(this, 'validate', { pushUpdatedManifests: config.pushUpdatedManifests });
    }
  }

  /**
   * Synthesizes all manifests into the output directory.
   */
  public synth(): void {
    if (!fs.existsSync(this.outdir)) {
      fs.mkdirSync(this.outdir);
    }

    Node.of(this).synthesize({
      outdir: this.outdir,
    });
  }
}
