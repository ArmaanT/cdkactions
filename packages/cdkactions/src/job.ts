import { Construct } from 'constructs';
import { StringMap, RunProps, DefaultsProps } from './types';
import { renameKeys } from './utils';
import { Workflow } from './workflow';


/**
 * Credentials to connect to a Docker registry with.
 */
export interface CredentialsProps {
  /**
   * Username to connect with.
   */
  readonly username: string;

  /**
   * Password to connect with.
   */
  readonly password: string;
}

/**
 * Generic Docker configuration.
 */
export interface DockerProps {
  /**
   * Image to use.
   */
  readonly image: string;

  /**
   * Credential configuration.
   */
  readonly credentials?: CredentialsProps;

  /**
   * Additional environment variables.
   */
  readonly env?: StringMap;

  /**
   * Ports to map.
   */
  readonly ports?: string[];

  /**
   * Volumes to map.
   */
  readonly volumes?: string[];

  /**
   * Additional Docker options.
   */
  readonly options?: string;
}

/**
 * Propsuration for a single GitHub Action step.
 */
export interface StepsProps extends RunProps {
  /**
   * A unique identifier.
   */
  readonly id?: string;

  /**
   * When to run this step.
   */
  readonly if?: string;

  /**
   * A name to display when running this action.
   */
  readonly name?: string;

  /**
   * Use an external action.
   */
  readonly uses?: string;

  /**
   * Commands to run.
   */
  readonly run?: string;

  /**
   * A map of parameters for an external action.
   */
  readonly with?: { [key: string]: string | number | boolean };

  /**
   * Additional environment variables.
   */
  readonly env?: StringMap;

  /**
   * Continue job if step fails.
   */
  readonly continueOnError?: boolean;

  /**
   * Maximum time before killing the step.
   */
  readonly timeoutMinutes?: number;
}

/**
 * Strategy configuration block.
 */
export interface StrategyProps {
  /**
   * A matrix to run jobs on.
   */
  readonly matrix?: { [key: string]: Array<any> };

  /**
   * Stop jobs when a single job fails.
   */
  readonly fastFail?: boolean;

  /**
   * Maximum parallel jobs.
   */
  readonly maxParallel?: number;
}

/**
 * Environment configuration block.
 */
export interface EnvironmentProps {
  /**
   * Name of the deployment environment.
   */
  readonly name?: string;

  /**
   * URL of the deployment environment.
   */
  readonly url?: string;
}

/**
 * Configuration for a single GitHub Action job.
 */
export interface JobProps {
  /**
   * Displayed name of the job.
   */
  readonly name?: string;

  /**
   * A job or list of jobs that must successfully complete before running this one.
   */
  readonly needs?: string | string[];

  /**
   * The type of machine to run on.
   */
  readonly runsOn: string;

  /**
   * A map of outputs for this job.
   */
  readonly outputs?: StringMap;

  /**
   * A map of environment variables to provide to the job.
   */
  readonly env?: StringMap;

  /**
   * The deployment environment of this job.
   */
  readonly environment?: string | EnvironmentProps;

  /**
   * A map of default settings to apply to all steps in this job.
   */
  readonly defaults?: DefaultsProps;

  /**
   * When to run this job.
   */
  readonly if?: string;

  /**
   * A list of steps to run.
   */
  readonly steps: StepsProps[];

  /**
   * Maximum time before killing the job.
   */
  readonly timeoutMinutes?: number;

  /**
   * A strategy configuration block.
   */
  readonly strategy?: StrategyProps;

  /**
   * Continue workflow if job fails.
   */
  readonly continueOnError?: boolean;

  /**
   * A container to run the job in.
   */
  readonly container?: DockerProps;

  /**
   * Additional Docker services provided to the job.
   */
  readonly services?: { [key: string]: DockerProps };
}

/**
 * Represents a GH Actions job.
 */
export class Job extends Construct {
  /**
   * An internal representation of a GH Action job
   */
  private readonly action: JobProps;

  /**
   * A unique identifier
   */
  public readonly id: string;

  /**
   * Defines a GitHub Actions job.
   * @param scope Workflow to run within.
   * @param id A unique identifier.
   * @param config The configuration of the job.
   */
  public constructor(scope: Workflow, id: string, config: JobProps) {
    super(scope, id);
    this.id = id;
    this.action = config;
  }

  /**
   * Converts the job's configuration into a format that is GitHub Actions compatible.
   */
  public toGHAction(): any {
    return renameKeys(this.action, {
      runsOn: 'runs-on',
      continueOnError: 'continue-on-error',
      timeoutMinutes: 'timeout-minutes',
      fastFail: 'fail-fast',
      maxParallel: 'max-parallel',
      workingDirectory: 'working-directory',
    });
  }
}
