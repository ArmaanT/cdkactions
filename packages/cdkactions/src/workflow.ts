import { Construct, Node } from 'constructs';
import { Job } from './job';
import { StringMap, DefaultsProps } from './types';
import { renameKeys, camelToSnake } from './utils';

/**
 * Configuration for the CheckRun event.
 */
export interface CheckRunTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'rerequested' | 'completed' | 'requested_action'>;
}

/**
 * Configuration for the CheckSuite event.
 */
export interface CheckSuiteTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'rerequested' | 'requested'>;
}

/**
 * Configuration for the IssueComment event.
 */
export interface IssueCommentTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the Issues event.
 */
export interface IssuesTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'opened' | 'edited' | 'deleted' | 'transferred' | 'pinned' | 'unpinned' | 'closed' | 'reopened' | 'assigned' | 'unassigned' | 'labeled' | 'unlabeled' | 'locked' | 'unlocked' | 'milestoned' | 'demilestoned'>;
}

/**
 * Configuration for the Label event.
 */
export interface LabelTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the Milestone event.
 */
export interface MilestoneTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'closed' | 'opened' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the ProjectTypes event.
 */
export interface ProjectTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'updated' | 'closed' | 'reopened' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the ProjectCard event.
 */
export interface ProjectCardTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'moved' | 'converted' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the ProjectColumn event.
 */
export interface ProjectColumnTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'updated' | 'moved' | 'deleted'>;
}

/**
 * Configuration for the PullRequest event.
 */
export interface PullRequestTypes extends PushTypes {
  /**
   * Supported types.
   */
  readonly types?: Array<'assigned' | 'unassigned' | 'labeled' | 'unlabeled' | 'opened' | 'edited' | 'closed' | 'reopened' | 'synchronize' | 'ready_for_review' | 'locked' | 'unlocked' | 'review_requested' | 'review_request_removed'>;
}

/**
 * Configuration for the PullRequestReview event.
 */
export interface PullRequestReviewTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'submitted' | 'edited' | 'dismissed'>;
}

/**
 * Configuration for the PullRequestReviewComment event.
 */
export interface PullRequestReviewCommentTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'created' | 'edited' | 'deleted'>;
}

/**
 * Configuration for the PullRequestTarget event.
 */
export interface PullRequestTargetTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'assigned' | 'unassigned' | 'labeled' | 'unlabeled' | 'opened' | 'edited' | 'closed' | 'reopened' | 'synchronize' | 'ready_for_review' | 'locked' | 'unlocked' | 'review_requested' | 'review_request_removed'>;
}

/**
 * Configuration for the RegistryPackage event.
 */
export interface RegistryPackageTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'published' | 'updated'>;
}

/**
 * Configuration for the Release event.
 */
export interface ReleaseTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'published' | 'unpublished' | 'created' | 'edited' | 'deleted' | 'prereleased' | 'released'>;
}

/**
 * Configuration for the Watch event.
 */
export interface WatchTypes {
  /**
   * Supported types.
   */
  readonly types: Array<'started'>;
}

/**
 * Configuration for the Push event.
 */
export interface PushTypes {
  /**
   * Branches to trigger the workflow on.
   */
  readonly branches?: string[];

  /**
   * Branches to ignore when triggering the workflow.
   */
  readonly branchesIgnore?: string[];

  /**
   * Tags to trigger the workflow on.
   */
  readonly tags?: string[];

  /**
   * Tags to ignore when triggering the workflow.
   */
  readonly tagsIgnore?: string[];

  /**
   * Paths to trigger the workflow on.
   */
  readonly paths?: string[];

  /**
   * Paths to ignore when triggering the workflow.
   */
  readonly pathsIgnore?: string[];
}

/**
 * Configuration for the Schedule event.
 */
export interface ScheduleEvent {
  /**
   * A cron schedule to run the workflow on.
   */
  readonly schedule: [{ 'cron': string }];
}

/**
 * Configuration for the WorkflowRun event.
 */
export interface WorkflowRunEvent {
  /**
   * A list of workflows to trigger from.
   */
  readonly workflows: string[];

  /**
   * Branches to trigger this workflow on.
   */
  readonly branches?: string[];

  /**
   * Branches to ignore when triggering this workflow.
   */
  readonly branchesIgnore?: string[];

  /**
   * Supported types.
   */
  readonly types?: Array<'completed' | 'requested'>;
}

/**
 * Events without additional subtypes.
 */
export type EventStrings = 'workflowDispatch' | 'repositoryDispatch' | 'create' | 'delete' | 'deployment' | 'deploymentStatus' | 'fork' | 'gollum' | 'pageBuild' | 'public' | 'status';

/**
 * Events with additional subtypes.
 */
export interface EventMap {
  /**
   * The checkRun event.
   */
  readonly checkRun?: CheckRunTypes;

  /**
   * The checkSuite event.
   */
  readonly checkSuite?: CheckSuiteTypes;

  /**
   * The issueComment event.
   */
  readonly issueComment?: IssueCommentTypes;

  /**
   * The issues event.
   */
  readonly issues?: IssuesTypes;

  /**
   * The label event.
   */
  readonly label?: LabelTypes;

  /**
   * The milestone event.
   */
  readonly milestone?: MilestoneTypes;

  /**
   * The project event.
   */
  readonly project?: ProjectTypes;

  /**
   * The projectCard event.
   */
  readonly projectCard?: ProjectCardTypes;

  /**
   * The projectColumn event.
   */
  readonly projectColumn?: ProjectColumnTypes;

  /**
   * The pullRequest event.
   */
  readonly pullRequest?: PullRequestTypes;

  /**
   * The pullRequestReview event.
   */
  readonly pullRequestReview?: PullRequestReviewTypes;

  /**
   * The pullRequestReviewComment event.
   */
  readonly pullRequestReviewComment?: PullRequestReviewCommentTypes;

  /**
   * The pullRequestTarget event.
   */
  readonly pullRequestTarget?: PullRequestTargetTypes;

  /**
   * The push event.
   */
  readonly push?: PushTypes;

  /**
   * The registryPackage event.
   */
  readonly registryPackage?: RegistryPackageTypes;

  /**
   * The release event.
   */
  readonly release?: ReleaseTypes;

  /**
   * The watch event.
   */
  readonly watch?: WatchTypes;
}

/**
 * All the events that can trigger a workflow.
 */
export type Events = keyof EventMap | EventStrings;

/**
 * Configuration for a single GitHub Action workflow.
 */
export interface WorkflowProps {
  /**
   * Name of the workflow.
   */
  readonly name: string;

  /**
   * When to run this workflow.
   */
  readonly on: Events | Array<Events> | EventMap | ScheduleEvent | WorkflowRunEvent;

  /**
   * A map of environment variables to provide to the job.
   */
  readonly env?: StringMap;

  /**
   * A map of default settings to apply to all steps in this job.
   */
  readonly defaults?: DefaultsProps;
}

/**
 * Represents a GH Action workflow.
 */
export class Workflow extends Construct {
  /**
   * File to save synthesized workflow manifest in.
   */
  public readonly outputFile: string;

  /**
   * An internal representation of a GH Action workflow.
   */
  private readonly action: WorkflowProps;

  /**
   * Represents a GitHub Actions workflow.
   * @param scope An ActionsStack instance.
   * @param id The name of this workflow.
   * @param action The configuration for this workflow.
   */
  public constructor(scope: Construct, id: string, config: WorkflowProps) {
    super(scope, id);
    this.action = config;
    this.outputFile = `cdkactions_${id}.yaml`;
  }

  /**
   * Converts the workflow's configuration into a format that is GitHub Actions compatible.
   */
  public toGHAction(): any {
    const workflow = renameKeys(this.action, {
      branchesIgnore: 'branches-ignore',
      tagsIgnore: 'tags-ignore',
      pathsIgnore: 'paths-ignore',
      checkRun: 'check_run',
      checkSuite: 'check_suite',
      issueComment: 'issue_comment',
      projectCard: 'project_card',
      projectColumn: 'project_column',
      pullRequest: 'pull_request',
      pullRequestReview: 'pull_request_review',
      pullRequestReviewComment: 'pull_request_review_comment',
      pullRequestTarget: 'pull_request_target',
      registryPackage: 'registry_package',
      workingDirectory: 'working-directory',
    });
    if (typeof workflow.on == 'string') {
      workflow.on = camelToSnake(workflow.on);
    }
    if (Array.isArray(workflow.on)) {
      workflow.on = workflow.on.map(camelToSnake);
    }
    const ghaction: any = { ...workflow, jobs: {} };
    for (const child of Node.of(this).children) {
      if (child instanceof Job) {
        // AWS constructs ensure children of constructs must have unique ids.
        ghaction.jobs[child.id] = child.toGHAction();
      }
    }
    return ghaction;
  }
}

