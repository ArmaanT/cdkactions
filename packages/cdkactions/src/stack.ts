import * as fs from 'fs';
import * as path from 'path';
import { Construct, ISynthesisSession, Node } from 'constructs';
import * as yaml from 'js-yaml';
import { Workflow } from './workflow';

/**
 * Represents a cdkaction stack.
 */
export class Stack extends Construct {

  /**
   * A unique identifier for the stack. Will be printed in synthesized manifests.
   */
  private readonly id: string;

  /**
   * Defines a cdkaction stack.
   * @param scope cdkaction app.
   * @param id A unique identifier for this stack.
   */
  public constructor(scope: Construct, id: string) {
    super(scope, id);
    this.id = id;
  }

  /**
   * A custom `onSynthesize` function to generate GH Action manifests.
   * @param session Synthesis session.
   */
  protected onSynthesize(session: ISynthesisSession) {
    for (const child of Node.of(this).children) {
      if (child instanceof Workflow) {
        const workflowOutput = path.join(session.outdir, child.outputFile);

        // Get raw yaml of workflow
        const rawYaml = yaml.dump(child.toGHAction(), { lineWidth: -1, noCompatMode: true, quotingType: '"' });

        // Save the formatted yaml
        fs.writeFileSync(workflowOutput, `# Generated by cdkactions. Do not modify\n# Generated as part of the '${this.id}' stack.\n`);
        fs.writeFileSync(workflowOutput, rawYaml, { flag: 'a' });
      }
    }
  }
}
