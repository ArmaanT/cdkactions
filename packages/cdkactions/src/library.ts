import { Construct } from 'constructs';
import dedent from 'ts-dedent';
import { Job, JobProps, StepsProps } from './job';
import { Stack } from './stack';
import { Workflow } from './workflow';


/**
 * Configuration for a CDKActionsStack instance.
 */
export interface CDKActionsProps {
  /**
   * Push updated manifests if they are out of date.
   * @default - false
   */
  readonly pushUpdatedManifests?: boolean;
}

/**
 * Provided CDKActions Stack that configures a workflow to validate cdkactions
 */
export class CDKActionsStack extends Stack {
  public constructor(scope: Construct, id: string, config: CDKActionsProps) {
    super(scope, id);
    const token = config.pushUpdatedManifests ? 'secrets.CDKACTIONS_TOKEN' : 'github.token';
    const synth = new Workflow(this, 'validate', {
      name: 'Validate cdkactions manifests',
      on: 'push',
    });
    new Job(synth, 'validate', {
      name: 'Validate cdkactions manifests',
      runsOn: 'ubuntu-latest',
      steps: [
        {
          uses: 'actions/checkout@v2',
          with: {
            token: `\${{ ${token} }}`,
          },
        },
        {
          name: 'Validate manifests',
          run: dedent`cd .github/cdk
                yarn install
                yarn build
                git --no-pager diff ../workflows
                git diff-index --quiet HEAD -- ../workflows`,
        },
        {
          name: 'Push updated manifests',
          if: config.pushUpdatedManifests ? 'always()' : 'false',
          run: dedent`cd .github/workflows
                git config user.name github-actions
                git config user.email github-actions[bot]@users.noreply.github.com
                git add .
                git commit -m "Update cdkactions manifests" || exit 0
                git push`,
        },
      ],
    });
  }
}

/**
 * A special Job that includes a checkout step automatically.
 */
export class CheckoutJob extends Job {
  public constructor(scope: Workflow, id: string, config: JobProps) {
    const steps: StepsProps[] = ([{ uses: 'actions/checkout@v2' }] as StepsProps[]).concat(config.steps);
    super(scope, id, { ...config, steps });
  }
}
