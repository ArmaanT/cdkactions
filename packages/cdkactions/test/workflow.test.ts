import { Construct } from 'constructs';
import { Job } from '../src';
import { TestingWorkflow } from './utils';

test('toGHAction', () => {
  const workflow = TestingWorkflow({
    name: 'Test',
    on: {
      pullRequest: {
        types: ['opened'],
        paths: ['test/'],
        pathsIgnore: ['ignore-path/'],
      },
      issueComment: { types: ['created'] },
    },
    defaults: {
      run: {
        workingDirectory: '~/',
      },
    },
    permissions: {
      'id-token': 'write',
      'contents': 'write',
    },
  });
  expect(workflow.toGHAction()).toMatchSnapshot();
});

test('on: string snake conversion', () => {
  const workflow = TestingWorkflow();
  expect(workflow.toGHAction().on).toBe('pull_request');
});

test('on: string[] snake conversion', () => {
  const workflow = TestingWorkflow({
    on: ['pullRequest', 'push', 'issueComment'],
  });
  const expected = ['pull_request', 'push', 'issue_comment'];
  expect(workflow.toGHAction().on).toEqual(expected);
});

test('2 jobs with same key -> error', () => {
  const workflow = TestingWorkflow();
  new Job(workflow, 'job', {
    runsOn: 'ubuntu-latest',
    steps: [],
  });
  expect(() =>
    new Job(workflow, 'job', {
      runsOn: 'ubuntu-latest',
      steps: [],
    }),
  ).toThrowError("There is already a Construct with name 'job' in Workflow [test]");
},
);

test('jobs kept in insertion order', () => {
  const workflow = TestingWorkflow();
  const job_one = 'job_one';
  const job_two = 'job_two';
  const job_three = 'job_three';
  new Job(workflow, job_one, {
    runsOn: 'ubuntu-latest',
    steps: [],
  });
  new Job(workflow, job_two, {
    runsOn: 'ubuntu-latest',
    steps: [],
  });
  new Job(workflow, job_three, {
    runsOn: 'ubuntu-latest',
    steps: [],
  });
  const jobs = Object.keys(workflow.toGHAction().jobs);
  const expected = [job_one, job_two, job_three];
  expect(jobs).toEqual(expected);
},
);

test('non-job children are ignored', () => {
  const workflow = TestingWorkflow();
  new Construct(workflow, 'not_job');
  expect(workflow.toGHAction().jobs).toEqual({});
});
