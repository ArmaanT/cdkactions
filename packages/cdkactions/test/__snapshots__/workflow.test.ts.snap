// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manual workflow dispatch 1`] = `
Object {
  "jobs": Object {},
  "name": "Test",
  "on": Object {
    "workflowDispatch": Object {
      "inputs": Object {
        "one": Object {
          "default": "foo",
          "description": "input one",
          "required": false,
          "type": "string",
        },
      },
    },
  },
}
`;

exports[`toGHAction 1`] = `
Object {
  "defaults": Object {
    "run": Object {
      "working-directory": "~/",
    },
  },
  "jobs": Object {},
  "name": "Test",
  "on": Object {
    "issue_comment": Object {
      "types": Array [
        "created",
      ],
    },
    "pull_request": Object {
      "paths": Array [
        "test/",
      ],
      "paths-ignore": Array [
        "ignore-path/",
      ],
      "types": Array [
        "opened",
      ],
    },
  },
}
`;
